File: softpos-cloud-auth-demo-android-java/app/src/main/java/it/nexi/softpos_evo_java_light/activities/ResultTransactionActivity.java
package it.nexi.softpos_evo_java_light.activities;


import static it.nexi.softpos_evo_java_light.domain.Constant.TYPE_OPERATION_ACCOUNTING_CLOSURE;
import static it.nexi.softpos_evo_java_light.domain.Constant.TYPE_OPERATION_LAST_TRANSACTION;
import static it.nexi.softpos_evo_java_light.domain.Constant.TYPE_OPERATION_PAYMENT;
import static it.nexi.softpos_evo_java_light.domain.Constant.TYPE_OPERATION_REVERSAL;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;

import it.nexi.softpos_evo_java_light.databinding.ActivityResultTransactionBinding;
import it.nexi.softpos_evo_java_light.domain.Constant;
import it.nexi.softpos_evo_java_light.domain.SoftPosResponseAccountingClosure;
import it.nexi.softpos_evo_java_light.domain.SoftPosResponseLastTransaction;
import it.nexi.softpos_evo_java_light.domain.SoftPosResponsePayment;
import it.nexi.softpos_evo_java_light.domain.SoftPosResponseReversal;
import it.nexi.softpos_evo_java_light.services.PreferencesService;
import it.nexi.softpos_evo_java_light.services.SDKNexiService;

/**
 *
 * Nexi Payment
 *
 * Receives via deeplink the outcome of the
 * transaction performed by Nexi POS
 *
 */
public class ResultTransactionActivity extends AppCompatActivity {

    private ActivityResultTransactionBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);

        binding = ActivityResultTransactionBinding.inflate(getLayoutInflater());

        setContentView(binding.getRoot());

        binding.btnCloseActivity.setOnClickListener(v -> finish());

    } // end protected void onCreate(

    // The handling of the transaction result from the Nexi POS
    // is managed within this method.
    // The Nexi POS sends the transaction result via a deeplink
    // to the response Uri specified in the intent call.
    @Override
    protected void onResume() {

        try {

            super.onResume();

            // Handle the transaction result sent by the Nexi POS
            if (getIntent() != null && getIntent().getData() != null
                    && getIntent().getData().getHost() != null){
                //  && (getIntent().getData().getHost().equals("payment"))){

                // Extracts the type of operation/method performed by the Nexi POS
                String method = getIntent().getData().getHost();

                // Extracts the data sent by Nexi POS
                Uri data = getIntent().getData();

                // Convert data to string
                String result = data.toString();

                // Format data string
                if (result != null && result.length() > 0){
                    result = result.replace("&","&\n");
                }

                // Show the result
                binding.editTextResultTransaction.setText(result);

                // Select the type of method/operation performed by the Nexi POS
                switch (method) {

                    case TYPE_OPERATION_PAYMENT:

                        // To perform a transaction reversal, the Reversal Activity must
                        // send specific parameters regarding the last executed payment
                        // transaction to the Nexi POS.
                        // Therefore, the necessary parameter values are saved in the
                        // app's preferences.

                        SoftPosResponsePayment softPosResponsePayment = SDKNexiService.parseSoftPosResponsePayment(data);

                        if (softPosResponsePayment != null) {

                            String terminalId = softPosResponsePayment.getTerminalID();

                            if (terminalId != null && !terminalId.isEmpty()) {

                                PreferencesService.setLastPaymentTerminalid(this, terminalId);

                            }

                            String totalAmount  = softPosResponsePayment.getTotalAmount();
                            String amount       = softPosResponsePayment.getAmount();
                            String tipAmount    = softPosResponsePayment.getTipAmount();

                            if (totalAmount != null && !totalAmount.isEmpty()) {

                                PreferencesService.setLastPaymentTotalAmount(this, totalAmount);
                                PreferencesService.setLastPaymentAmount(this, amount);
                                PreferencesService.setLastPaymentTipAmount(this, tipAmount);

                            }

                        } // end if (softPosResponsePayment != null)

                        break;

                    case TYPE_OPERATION_REVERSAL:

                        SoftPosResponseReversal softPosResponseReversal = SDKNexiService.parseSoftPosResponseReversal(data);

                        if (softPosResponseReversal != null) {

                            // ------

                        } // end if (softPosResponseLastTransaction != null)

                        break;

                    case TYPE_OPERATION_LAST_TRANSACTION:

                        // To perform a transaction reversal, the Reversal Activity must
                        // send specific parameters regarding the last executed payment
                        // transaction to the Nexi POS.
                        // Therefore, the necessary parameter values are saved in the
                        // app's preferences.

                        SoftPosResponseLastTransaction softPosResponseLastTransaction = SDKNexiService.parseSoftPosResponseGetLastTransaction(data);

                        if (softPosResponseLastTransaction != null) {

                            String terminalId = softPosResponseLastTransaction.getTerminalID();

                            if (terminalId != null && !terminalId.isEmpty()) {

                                PreferencesService.setLastPaymentTerminalid(this, terminalId);
                            }

                            String totalAmount = softPosResponseLastTransaction.getTotalAmount();
                            String amount = softPosResponseLastTransaction.getAmount();
                            String tipAmount = softPosResponseLastTransaction.getTipAmount();

                            if (totalAmount != null && !totalAmount.isEmpty()) {

                                PreferencesService.setLastPaymentTotalAmount(this, totalAmount);
                                PreferencesService.setLastPaymentAmount(this, amount);
                                PreferencesService.setLastPaymentTipAmount(this, tipAmount);
                            }

                        } // end if (softPosResponseLastTransaction != null)

                        break;

                    case TYPE_OPERATION_ACCOUNTING_CLOSURE:

                        SoftPosResponseAccountingClosure softPosResponseAccountingClosure  = SDKNexiService.parseSoftPosResponseAccountingClosure(data);

                        if (softPosResponseAccountingClosure != null) {

                            // ------

                        } // end if (softPosResponseLastTransaction != null)

                        break;

                    default:

                        Log.w(Constant.TAG_SOFTPOS_EVO_JAVA, "unknown method : " + method);

                        break;

                } // end switch (host)

            } else {

                Log.e(Constant.TAG_SOFTPOS_EVO_JAVA, "Server error");

            } // end if (getIntent() != null && getIntent().getData() != null

        } catch (Exception e) {

            binding.editTextResultTransaction.setText(e.getMessage());

            Log.e(Constant.TAG_SOFTPOS_EVO_JAVA, "Deeplink parsing error: " + e.getMessage());
        }

    } // end  protected void onResume() {

} // end Activity

File: softpos-cloud-auth-demo-android-kotlin/app/src/main/java/it/nexi/softpos_evo_kotlin_01/activities/ResultTransactionActivity.kt
package it.nexi.softpos_evo_kotlin_01.activities

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import android.net.Uri
import it.nexi.softpos_evo_kotlin_01.databinding.ActivityResultTransactionBinding
import it.nexi.softpos_evo_kotlin_01.domain.Constant.TYPE_OPERATION_ACCOUNTING_CLOSURE
import it.nexi.softpos_evo_kotlin_01.domain.Constant.TYPE_OPERATION_LAST_TRANSACTION
import it.nexi.softpos_evo_kotlin_01.domain.Constant.TYPE_OPERATION_PAYMENT
import it.nexi.softpos_evo_kotlin_01.domain.Constant.TYPE_OPERATION_REVERSAL
import it.nexi.softpos_evo_kotlin_01.services.PreferencesServvice

/**
 *
 * Nexi Payment
 *
 * Receives via deeplink the outcome of the
 * transaction performed by Nexi POS
 *
 */

class ResultTransactionActivity : AppCompatActivity() {

    private lateinit var binding: ActivityResultTransactionBinding

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)

        binding = ActivityResultTransactionBinding.inflate(layoutInflater)

        setContentView(binding.root)

        binding.btnCloseActivity.setOnClickListener {

            finish()

        }

    } // onCreate

    // The handling of the transaction result from the Nexi POS
    // is managed within this method.
    // The Nexi POS sends the transaction result via a deeplink
    // to the response Uri specified in the intent call.
    override fun onResume() {

        super.onResume()

        val uri = intent?.data ?: return
        val operation = uri.host ?: return

        try {

            val resultString = uri.toString()

            Log.d("softpos_kotlin", "Operation: $operation, URI: $resultString")

            binding.editTextResultTransaction.setText(resultString)

            // Select the type of method/operation performed by the Nexi POS
            when (operation) {

                TYPE_OPERATION_PAYMENT -> {

                    parseSoftPosResponsePayment(uri)

                }

                TYPE_OPERATION_REVERSAL -> {

                }

                TYPE_OPERATION_LAST_TRANSACTION -> {

                }

                TYPE_OPERATION_ACCOUNTING_CLOSURE -> {

                }

                else -> {

                    Log.w("softpos_kotlin", "Unknown operation: $operation")

                }

            } // when (operation) {


            intent.data = null

        } catch (e: Exception) {

            binding.editTextResultTransaction.setText(e.message)

        }

    } // onResume(

    /**
     *
     * Example of parsing the deeplink sent by Nexipos
     * relating to a payment transaction
     *
     */
    fun parseSoftPosResponsePayment(uri : Uri) {

        val callerTrxId: String?            = uri.getQueryParameter("callerTrxId")
        val operationType: String?          = uri.getQueryParameter("operationType")
        val result: String?                 = uri.getQueryParameter("result")
        val amount: String?                 = uri.getQueryParameter("amount")
        val actionCode: String?             = uri.getQueryParameter("actionCode")
        val pan: String?                    = uri.getQueryParameter("pan")
        val transactionType: String?        = uri.getQueryParameter("transactionType")
        val authorizationNumber: String?    = uri.getQueryParameter("authorizationNumber")
        val timeStamp: String?              = uri.getQueryParameter("timestamp")
        val resultDescription: String?      = uri.getQueryParameter("resultDescription")
        val cardTypeCVM: String?            = uri.getQueryParameter("cardTypeCVM")
        val acquireId: String?              = uri.getQueryParameter("acquirerId")
        val stan: String?                   = uri.getQueryParameter("stan")
        val operationNumber: String?        = uri.getQueryParameter("operationNumber")
        val acquirerName: String?           = uri.getQueryParameter("acquirerName")
        val terminalID: String?             = uri.getQueryParameter("terminalId")
        val merchantId: String?             = uri.getQueryParameter("merchantId")
        val urlTicket: String?              = uri.getQueryParameter("urlTicket")
        val autoClose: String?              = uri.getQueryParameter("autoClose")

        val amountInt = amount?.toIntOrNull()

        PreferencesServvice.savePaymentData(this,amountInt,terminalID)

    } // fun parseSoftPosResponsePayment

} // end activity